### ISSUE TITLE
migration vue-router 0.7 data guard suggestion

### ISSUE THREAD

**Issue by doggerFrankXu** [vuejs/vuejs.org#825]
For the [data migration](https://vuejs.org/v2/guide/migration-vue-router.html#data-replaced), it might be better using the`next` callback in `beforeRouteEnter` to replace `data` for 2 reasons.
First it can be triggered in create time, but `watch` cannot. (actually, I haven't know the details of `watch`.. I don't know where it placed in life-cycle diagram.)
Second it has similar parameter like `data` hook. I tried the [data fetching](http://router.vuejs.org/en/advanced/data-fetching.html) section's way but in `created` hook, I cannot get `from` .
   I am newbie to Vue but I am working on migration. And the reason I suggest this is that I have a keep-alive component and its `data` hook need `from` and `this`

**Comment by chrisvfritz** [comment]
Good point! I'll add a note that `beforeRouteEnter` may also be needed.
